From 02ad81160b685a7c0a054f7eedd3400854194e14 Mon Sep 17 00:00:00 2001
From: Colin Godsey <crgodsey@gmail.com>
Date: Mon, 8 Jul 2019 08:54:47 -0600
Subject: [PATCH] Bungee message for player ping


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 1f8a2439..6a39250e 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -483,6 +483,13 @@ public class DownstreamBridge extends PacketHandler
                     player.disconnect( new TextComponent( kickReason ) );
                 }
             }
+            // Waterfall start
+            if ( subChannel.equals( "GetPing" ) )
+            {
+                out.writeUTF( "GetPing" );
+                out.writeInt( con.getPing() );
+            }
+            // Waterfall end
 
             // Check we haven't set out to null, and we have written data, if so reply back back along the BungeeCord channel
             if ( out != null )
-- 
2.17.2 (Apple Git-113)

