From aae95890f526f48e6fa0d5bf266acc07e52ac950 Mon Sep 17 00:00:00 2001
From: Colin Godsey <crgodsey@gmail.com>
Date: Wed, 17 Apr 2019 06:51:16 -0600
Subject: [PATCH] ServerChannelTapEvent


diff --git a/api/src/main/java/network/ycc/waterdog/api/event/ServerChannelTapEvent.java b/api/src/main/java/network/ycc/waterdog/api/event/ServerChannelTapEvent.java
new file mode 100644
index 00000000..a1475d4a
--- /dev/null
+++ b/api/src/main/java/network/ycc/waterdog/api/event/ServerChannelTapEvent.java
@@ -0,0 +1,22 @@
+package network.ycc.waterdog.api.event;
+
+import io.netty.channel.Channel;
+
+import lombok.AllArgsConstructor;
+import lombok.EqualsAndHashCode;
+
+import lombok.Getter;
+import net.md_5.bungee.api.config.ServerInfo;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.api.plugin.Event;
+
+@AllArgsConstructor
+@EqualsAndHashCode(callSuper = false)
+public class ServerChannelTapEvent extends Event {
+    @Getter
+    private ProxiedPlayer user;
+    @Getter
+    private ServerInfo serverInfo;
+    @Getter
+    private Channel channel;
+}
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 7e4cdf38..8379a8fe 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -383,6 +383,7 @@ public final class UserConnection implements ProxiedPlayer
                         network.ycc.waterdog.pe.PEPipelineUtils.moveTimeout(ch);
                     }
                 }
+                bungee.getPluginManager().callEvent(new network.ycc.waterdog.api.event.ServerChannelTapEvent(UserConnection.this, request.getTarget(), ch));
                 // Waterdog end
                 PipelineUtils.BASE.initChannel( ch );
                 ch.pipeline().addAfter( PipelineUtils.FRAME_DECODER, PipelineUtils.PACKET_DECODER, new MinecraftDecoder( Protocol.HANDSHAKE, false, getPendingConnection().getVersion() ) );
-- 
2.17.2 (Apple Git-113)

