From 9323255244d457fa7ff554f6fd6b3cf2ec0eabb6 Mon Sep 17 00:00:00 2001
From: "beckerluca16@gmail.com" <Wasserkocher1998>
Date: Fri, 19 Jul 2019 14:42:53 +0200
Subject: [PATCH] Stop the change dimension sound


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index 7ebac7be..55eeba35 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -109,6 +109,10 @@ public enum Protocol
                     network.ycc.waterdog.pe.packet.PELocalPlayerInitialized.class,
                     map( ProtocolConstants.MINECRAFT_PE_1_8, 0x71 )
             );
+            TO_CLIENT.registerPacket(
+                    network.ycc.waterdog.pe.packet.PEStopSoundPacket.class,
+                    Protocol.map( ProtocolConstants.MINECRAFT_PE_1_8, 0x57 )
+            );
             // Waterdog end
             TO_CLIENT.registerPacket(
                     KeepAlive.class,
diff --git a/protocol/src/main/java/network/ycc/waterdog/pe/packet/PEStopSoundPacket.java b/protocol/src/main/java/network/ycc/waterdog/pe/packet/PEStopSoundPacket.java
new file mode 100644
index 00000000..dab227a1
--- /dev/null
+++ b/protocol/src/main/java/network/ycc/waterdog/pe/packet/PEStopSoundPacket.java
@@ -0,0 +1,35 @@
+package network.ycc.waterdog.pe.packet;
+
+import io.netty.buffer.ByteBuf;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+import net.md_5.bungee.protocol.AbstractPacketHandler;
+import net.md_5.bungee.protocol.DefinedPacket;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@EqualsAndHashCode(callSuper = false)
+public class PEStopSoundPacket extends DefinedPacket {
+
+    public String name;
+    public boolean stopAll;
+
+    @Override
+    public void read( ByteBuf buf ) {
+        //Ignore
+    }
+
+    @Override
+    public void write( ByteBuf buf ) {
+        writeString( name, buf );
+        buf.writeBoolean( stopAll );
+    }
+
+    @Override
+    public void handle( AbstractPacketHandler handler ) throws Exception {
+
+    }
+}
\ No newline at end of file
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 7bfeb183..67782804 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -64,6 +64,7 @@ import net.md_5.bungee.tab.ServerUnique;
 import net.md_5.bungee.tab.TabList;
 import net.md_5.bungee.util.CaseInsensitiveSet;
 import net.md_5.bungee.util.ChatComponentTransformer;
+import network.ycc.waterdog.pe.packet.PEStopSoundPacket;
 
 @RequiredArgsConstructor
 public final class UserConnection implements ProxiedPlayer
@@ -170,6 +171,7 @@ public final class UserConnection implements ProxiedPlayer
                     net.md_5.bungee.protocol.ProtocolConstants.isPE(getPendingConnection().getVersion())) {
                 // TODO: find a better place for this
                 network.ycc.waterdog.pe.PERawPacketData.injectForcedDimChange(ch.getHandle(), getPendingConnection().getVersion());
+                unsafe.sendPacket( new PEStopSoundPacket( "portal.travel", true ) );
             }
             // Waterdog end
         }
-- 
2.20.1.windows.1

