From f66e4685184ad4adbcb7490890f87743812e0c65 Mon Sep 17 00:00:00 2001
From: Alemiz112 <45739737+Alemiz112@users.noreply.github.com>
Date: Fri, 6 Dec 2019 20:27:45 +0100
Subject: [PATCH] Waterdog PE (script modifications)


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
index d9497970..883d9945 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
@@ -58,7 +58,12 @@ public class BungeeCordLauncher
 
         BungeeCord bungee = new BungeeCord();
         ProxyServer.setInstance( bungee );
-        bungee.getLogger().info( "Enabled Waterfall version " + bungee.getVersion() );
+        bungee.getLogger().info( ChatColor.AQUA + "Starting "+VersionInfo.SOFTWARE+" proxy software!");
+        bungee.getLogger().info( ChatColor.DARK_AQUA + "Software Version: "+VersionInfo.VERSION);
+        bungee.getLogger().info( ChatColor.DARK_AQUA + "Build Version: "+VersionInfo.JENKINS_BUILD_ID);
+        bungee.getLogger().info( ChatColor.DARK_AQUA + "Raw Version: "+bungee.getVersion());
+        bungee.getLogger().info( ChatColor.DARK_AQUA + "Development Build: "+VersionInfo.IS_DEVELOPMENT_BUILD);
+        bungee.getLogger().info( ChatColor.DARK_AQUA + "Software Authors: "+VersionInfo.AUTHORS);
         bungee.start();
 
         if ( !options.has( "noconsole" ) )
diff --git a/proxy/src/main/java/net/md_5/bungee/VersionInfo.java b/proxy/src/main/java/net/md_5/bungee/VersionInfo.java
new file mode 100644
index 00000000..6a15b9c2
--- /dev/null
+++ b/proxy/src/main/java/net/md_5/bungee/VersionInfo.java
@@ -0,0 +1,10 @@
+package net.md_5.bungee;
+
+public interface VersionInfo {
+
+    public static String SOFTWARE = "WaterDog";
+    public static String VERSION = "1.0.0";
+    public static String JENKINS_BUILD_ID = "unknown";
+    public static boolean IS_DEVELOPMENT_BUILD = false;
+    public static String AUTHORS = "YesDogOSS Team, Bungee Team";
+}
-- 
2.23.0.windows.1

